### PR TITLE
Add traits SLG solver section

### PR DESCRIPTION
**Pull request by tmandry**
This is mostly just an adapted version of Niko's blog post.

From here we need to add a subsection on negative reasoning, and possibly one on how the solver handles regions. Also one on using the solver via the `Context` and related traits.

cc @nikomatsakis @scalexm

### CODE REVIEW

**Comment by scalexm** [comment]
I don't have any objections.

File: src/traits/slg.md [hunk 1]
```diff
@@ -1,3 +1,283 @@
...
+    A0 = [?0 = u32]
+  Strand:
+    (Vec<?U>: Debug) :- (?U: Debug)
+    (Rc<?V>: Debug) :- (?V: Debug)
+```
+
+Note that I am writing out the answer A0 as a substitution that can be
+applied to the table goal; actually, in the code, the goals for each
+X-clause are also represented as substitutions, but in this exposition
+I've chosen to write them as full goals, following NFTD.
```
Comment (mark-i-m) [inline src/traits/slg.md]: NFTD?

File: src/traits/slg.md [hunk 2]
```diff
@@ -1,3 +1,283 @@
...
+   
+Since we now have an answer, `ensure_answer(T1, A0)` will return `Ok`
+to the table T0, indicating that answer A0 is available. T0 now has
+the job of incorporating that result into its active strand. It does
+this in two ways. First, it creates a new strand that is looking for
+the next possible answer of T1. Next, it incorpoates the answer from
```
Comment (mark-i-m) [inline src/traits/slg.md]: We create this new strand so that if anyone asks for `T0, A1`, we have a strand that can compute it, right?

Comment (tmandry) [reply src/traits/slg.md]: Yes. In particular, the new strand "remembers" that more answers from `T1` (starting with `T1, A1` should be incorporated into `T0`.

**Comment by tmandry** [comment]
Waiting on https://github.com/rust-lang-nursery/chalk/pull/190 to be merged so I can link to some methods.

**Comment by mark-i-m** [comment]
@tmandry r=me :) Should I wait for Niko?

**Pull request merged**